src/block_ffm.rs

```diff
@@ -398,177 +316,228 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
-                            for k in 0..FFMK {
-                                myslice[f1 * fb.ffm_fields_count as usize + f2] += contra_fields
-                                    .get_unchecked(f1_offset_ffmk + k as usize)
-                                    * contra_fields.get_unchecked(f2_offset_ffmk + k as usize)
-                                    * 0.5;
-                                myslice[f2 * fb.ffm_fields_count as usize + f1] += contra_fields
-                                    .get_unchecked(f1_offset_ffmk + k as usize)
-                                    * contra_fields.get_unchecked(f2_offset_ffmk + k as usize)
-                                    * 0.5;
-                            }
-                        }
+                    let mut correction = 0.0;
+                    for k in feature_field_index..feature_field_index + ffmk_as_usize {
+                        correction += ffm_weights.get_unchecked(k) * ffm_weights.get_unchecked(k);
                     }
-                });
-            } else {
-                // Old straight-forward method. As soon as we have multiple feature values per field, it is slower
-                let ffm_weights = &self.weights;
-                specialize_k!(self.ffm_k, FFMK, wsumbuf, {
-                    for (i, left_hash) in fb.ffm_buffer.iter().enumerate() {
-                        for right_hash in fb.ffm_buffer.get_unchecked(i + 1..).iter() {
-                            //if left_hash.contra_field_index == right_hash.contra_field_index {
-                            //    continue	// not combining within a field
-                            //}
-                            let joint_value = left_hash.value * right_hash.value;
-                            let lindex = left_hash.hash + right_hash.contra_field_index;
-                            let rindex = right_hash.hash + left_hash.contra_field_index;
-                            for k in 0..FFMK {
-                                let left_hash_weight =
-                                    ffm_weights.get_unchecked((lindex + k) as usize).weight;
-                                let right_hash_weight =
-                                    ffm_weights.get_unchecked((rindex + k) as usize).weight;
-                                *wsumbuf.get_unchecked_mut(k as usize) +=
-                                    left_hash_weight * right_hash_weight * joint_value;
+
+                    *myslice.get_unchecked_mut(((feature.contra_field_index / ffmk) * ffm_fields_count_plus_one) as usize) -=
+                        correction * 0.5 * feature_value * feature_value;
+
+                    ffm_buffer_index += 1;
+                    feature_num += 1;
+                }
+            }
+
+            let mut f1_offset = 0;
+            let mut f1_index_offset = 0;
+            let mut f1_ffmk = 0;
+            let mut self_interaction_index = 0;
+            for f1 in 0..ffm_fields_count_as_usize {
```

Comment:
`f1` and `f2` seem odd variable names, but ok

Reply:
I agree, but I left them the same as before